File: rayz/canvas.py
# import rayz.util as U
from rayz.color import *
from rayz.util import *

class Canvas:
    def __init__(self, height:int=0, width:int=0)  -> None:
        self.width = width
        self.height = height
        self.grid = [[Color(red=0, green=0, blue=0) for _ in range(self.width)] for _ in reversed(range(self.height))]

    def __str__(self) -> str:
        result = ""
        for y in range(self.height):
            for x in range(self.width):
                result += f"[{x},{y}]-{self.pixel_at(x=x, y=y)} "

            result += "\n"

        return result

    def pixel_at(self, x:int, y:int) -> Color:
        if x > self.width - 1:
          raise ColumnIndexTooLarge(f"Column index({x}) greater than {self.width - 1}")

        if y > self.height - 1:
          raise RowIndexTooLarge(f"Row index({y}) greater than {self.height - 1}")

        return self.grid[y][x]

    def write_pixel(self, x:int, y:int, color:Color) -> None:
        if x > self.width - 1:
          raise ColumnIndexTooLarge(f"Column index({x}) greater than {self.width - 1}")

        if y > self.height - 1:
          raise RowIndexTooLarge(f"Row index({y}) greater than {self.height - 1}")

        self.grid[y][x] = color

    def to_ppm(self) -> str:
        def clamp(value: float, min: int = 0, max: int = 255) -> int:
            n = round(value * max)
            return max if n > max else min if n < min else n

        ppm = f"P3\n{self.width} {self.height}\n255\n"
        for row in self.grid:
            for color in row:
                ppm  = ppm + f"{clamp(color.red)} {clamp(color.green)} {clamp(color.blue)} "
            ppm = ppm.rstrip()
            ppm = ppm + "\n"

        ppm = ppm + "\n"
        return ppm

File: rayz/lib/canvas.rb
module Rayz
  module Lib
    class Canvas
      attr_reader :height, :width, :pixels
      MAX_COLORS = 255

      def initialize(width:, height:)
        @width = width
        @height = height
        black = Color.new(red: 0.0, green: 0.0, blue: 0.0)
        @pixels = Array.new(height) { Array.new(width, black) }
      end

      def to_ppm
        ppm_header + ppm_body + ppm_footer
      end

      def write_pixel(row:, col:, color:)
        raise "write_pixel: Row out of bounds: #{row}" if row > @height - 1 || row < 0
        raise "write_pixel: Col out of bounds: #{col}" if col > @width - 1 || col < 0

        @pixels[row][col] = color
      end

      def pixel_at(row:, col:)
        raise "pixel_at: Row out of bounds: #{row}" if row > @height - 1 || row < 0
        raise "pixel_at: Col out of bounds: #{col}" if col > @width - 1 || col < 0

        @pixels[row][col]
      end

      def to_s
        output = "Height: #{@height}, Columns: #{@width}\n\n"

        (0..@height - 1).to_a.reverse_each do |r|
          (0..@width - 1).to_a.each do |c|
            output << "[R:#{r} C:#{c}] - #{@pixels[r][c]} "
          end
          output << "\n"
        end
        output
      end

      private

      def ppm_header
        <<~HEREDOC
          P3
          #{@width} #{@height}
          #{MAX_COLORS}
        HEREDOC
      end

      def ppm_body
        output = [] # : Array[String]
        total_values = (MAX_COLORS + 1)

        (0..@height - 1).to_a.reverse_each do |r|
          chunk = [] # : Array[String]
          (0..@width - 1).each do |c|
            scaled_red = (@pixels[r][c].red * total_values).round
            clamped_red = scaled_red.clamp(0, MAX_COLORS)

            scaled_green = (@pixels[r][c].green * total_values).round
            clamped_green = scaled_green.clamp(0, 255)

            scaled_blue = (@pixels[r][c].blue * total_values).round
            clamped_blue = scaled_blue.clamp(0, 255)

            chunk << "#{clamped_red} #{clamped_green} #{clamped_blue} "
          end
          output << chunk.reverse.join.rstrip + "\n"
        end
        output.join
      end

      def ppm_footer
        "\n"
      end
    end
  end
end
